src/lib.rs

```diff
@@ -1,7 +1,35 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
-    }
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! rust-vmm device model.
+
+extern crate vm_memory;
+
+use vm_memory::GuestAddress;
+
+/// IO Resource type.
+#[derive(Debug, Copy, Clone)]
+pub enum IoType {
+    /// Port I/O type.
+    Pio,
+
+    /// Memory mapped I/O type.
+    Mmio,
+}
+
+/// Device IO trait.
+/// A device supporting memory based I/O should implement this trait, then
+/// register itself against the different IO type ranges it handles.
+/// The VMM will then dispatch IO (PIO or MMIO) VM exits by calling into the
+/// registered devices read or write method from this trait.
+#[allow(unused_variables)]
+pub trait DeviceIo: Send {
+    /// Get the device name.
+    fn name(&self) -> String;
+
+    /// Read from the guest physical address `addr` to `data`.
+    fn read(&mut self, addr: GuestAddress, data: &mut [u8], io_type: IoType);
```

Comment:
I'm thinking about Arc<Mutex<dyn Device>> vs Arc<dyn Device>. I feel it would more flexible to use Arc<dyn Device> and lety the implementation to protect itself from concurrent accesses on demand.
If that's the preferred way, we need to use "&self" instead of "&mut self" here.